### PR TITLE
added Hong Kong Observatory as a new source

### PR DESCRIPTION
**Pull request by chunshek**
I have added Hong Kong Observatory as a new official source for Hong Kong. I have implemented the following:
- Current Observations
- Monthly Normals
- Hourly/Daily Forecasts
- Warnings

The source can handle all three languages in which HKO publishes its forecast and warning text: English, Traditional Chinese, Simplified Chinese.

I will also create a Chinese (Hong Kong) translation for the app on Weblate. It will be initially based on the existing Traditional Chinese `zh-TW` translation, but revised with certain Hong Kong terminology to make the app interface more consistent with the source output.

BTW, the HKO has official translations for its weather conditions and warning titles in eight additional languages. Those are all prefixed with `hko_weather_text_` and `hko_warning_text_` in `strings.xml`. Once the file has been updated on Weblate, I can update those translations accordingly where applicable. The eight additional languages are:
- Bahasa Indonesia
- Hindi
- Nepali
- Punjabi
- Tagalog
- Thai
- Urdu
- Vietnamese

### CODE REVIEW

**Comment by papjul** [comment]
Thank you!

> I will also create a Chinese (Hong Kong) translation for the app on Weblate. It will be initially based on the existing Traditional Chinese `zh-TW` translation, but revised with certain Hong Kong terminology to make the app interface more consistent with the source output.

Why is that required? For the history, initially we had split Hong Kong/Taiwan translation files, but they were identical, so I moved them to Simplified/Traditional only, due to many translator conflicts and lack of maintenance of both traditional files.

> BTW, the HKO has official translations for its weather conditions and warning titles in eight additional languages. Those are all prefixed with `hko_weather_text_` and `hko_warning_text_` in `strings.xml`. Once the file has been updated on Weblate, I can update those translations accordingly where applicable. The eight additional languages are:
> 
>     * Bahasa Indonesia
> 
>     * Hindi
> 
>     * Nepali
> 
>     * Punjabi
> 
>     * Tagalog
> 
>     * Thai
> 
>     * Urdu
> 
>     * Vietnamese

I'm a bit reluctant to add so many translations for a very specific source in Breezy Weather itself, can’t they be fetched from the API (even if it requires an additional call)? That way, if other languages are added, they will automatically be supported.

**Comment by chunshek** [comment]
### On the need for separate translations for `zh-TW` and `zh-HK`
I happen to be a Hongkonger living in Taiwan, so am acutely aware of some key differences in terminology. Here are a few examples (not exhaustive):
| English | Chinese (Taiwan) | Chinese (Hong Kong) |
|---|---|---|
| Air Quality | 空氣品質 | 空氣質素 |
| Alert | 警特報 | 警報 |
| cm _(centimeter)_ | 公分 | 厘米 |
| m _(meter)_ | 公尺 | 米 |
| UV very high | 過量級 | 甚高 |
| UV extreme | 危險級 | 極高 |
| Relative humidity | 相對溼度 | 相對濕度 |
| Partly cloudy | **多雲** | 部分時間多雲 |
| Cloudy | 陰 | **多雲** |
| Overcast | 天陰 | 密雲 |
| Rain showers | 陣雨 | 驟雨 |
| Haze | 霾 | 煙霞 |
| _(all 13 Beaufort forces)_ | _(all using CWA terms)_ | _(all using HKO terms)_ |

I would agree that the UI phrasing between the two should be mostly the same (or with minor differences that are understandable). However specific weather conditions are described in marked different terminology between the two places, enough to cause confusion if used interchangeably – especially with 多雲 meaning **partly cloudy** in Taiwan, but **"fully" cloudy** in Hong Kong!

Another example: HKO groups both Beaufort forces 1 and 2 under 輕微 (literally "light and minor"), but in Taiwan 輕風 ("light wind") is force 2 and 微風 ("minor wind") is already force 3. A Hongkonger seeing 微風 in `zh-TW` may be led to think the wind is softer than it actually is (though force 3 wind is not hazardous at all, but you get my point).

If the app can have English, English (US), English (UK), English (Canada), English (Australia) translations whose sole purposes are to serve spelling differences, it stands to separate Chinese (Taiwan) and Chinese (Hong Kong) to serve actual semantic differences.

### On minor languages
I wouldn't suggest adding new translations to Breezy Weather. I was simply saying that once the main `strings.xml` is updated on Weblate, I was gonna go and fill in the official HKO strings for the ones whose translations already exist. That means only for:
- Bahasa Indonesia _(Indonesian)_
- Hindi
- Vietnamese

Since there are no translations for Nepali, Punjabi, Tagalog _(Filipino)_, Thai and Urdu right now, I'll skip those for sure.

Unfortunately we can't pull those with an API call, as those strings are hardcoded in some very ugly Javascript files. (Not minified, but still pretty atrocious as far as codes go.) I don't think any staff at HKO speaks any of those languages. They probably just hired translators for one-off jobs to build a set of largely automated sites to serve linguistic minorities.

**Comment by papjul** [comment]
> If the app can have English, English (US), English (UK), English (Canada), English (Australia) translations whose sole purposes are to serve spelling differences, it stands to separate Chinese (Taiwan) and Chinese (Hong Kong) to serve actual semantic differences.

There's a big difference between English and Chinese. In English, ONLY the changed strings are re-translated in regional variants, that's only a few strings, the rest is inherited from the default translation file. Btw, the spelling is just a bonus, the regional variants were created mostly to handle date formatting differences.

The issue in Chinese (or with any other language I think) is that you need to duplicate all translations, and there are a bunch of them.

So I understand how important it is to have different regional translations thanks to your example, but there are still a few issues:
- Most of the time, translators are the same for CN, HK and TW. Do they know all the differences when they translate?
- What if some translators don't agree between each other due to regional differences? The Chinese languages are amongst the most active on Weblate, and I can't decide who is right as I don't speak Chinese.

There are not that kind of issues in portuguese which is in a similar case. I also think a latin language is probably less prone to major debates.

> I wouldn't suggest adding new translations to Breezy Weather.

If I merge your PR, your added translations will be pushed to Weblate and all translators will be notified they need to translate them.

As a workaround, you can make them all non translatable (I saw you did it for a few) but that means all translations will need to be maintained on GitHub and not through Weblate. And Android Studio won't be happy, but it works.

**Comment by chunshek** [comment]
**TL;DR:**
1. Chinese translations don't have to "agree" between variants.
2. Fine with removing HKO specific text from `strings.xml` and hardcoding them if that's what you prefer.

### More thoughts about Chinese translations for Breezy Weather

> There's a big difference between English and Chinese. In English, ONLY the changed strings are re-translated in regional variants, that's only a few strings, the rest is inherited from the default translation file.

> The issue in Chinese (or with any other language I think) is that you need to duplicate all translations, and there are a bunch of them.

I know that. I wish it were that simple for Chinese as well. :-/

> - Most of the time, translators are the same for CN, HK and TW. Do they know all the differences when they translate?
> - What if some translators don't agree between each other due to regional differences? The Chinese languages are amongst the most active on Weblate, and I can't decide who is right as I don't speak Chinese.

That's the thing: they don't know! I was going through the `zh-TW` translation for Breezy Weather a while back, and saw in there many instances of simple conversions of `zh-CN` strings from Simplified characters to Traditional. They clearly did not fit Taiwanese usage and looked especially out of place with the CWA source I contributed. So I revised a bunch of them.

For L10N, the differences between `zh-CN` and `zh-TW` are much more numerable and profound than, say between `en-US` and `en-GB`, or `fr-FR` and `fr-CA`. To me, they really ought to be treated as separate languages altogether. That means translators for `zh-CN` and `zh-TW` should **not** agree with each other. It's natural that they don't. Otherwise there would be no need for separate variants in the first place. Of course, one variant can be used as the basis to start another variants, but it's fine that they split into their own eventually.

Long story short: You don't have to decide "who's right" between variants. You can just let them be different.

For my part, I am native in `zh-HK` and can proficiently revise and verify `zh-TW` translations as well, since I have lived here for a long time. Even though I can read `zh-CN` perfectly and have to use it a lot in my professional life, I won't be touching `zh-CN` translations. I can write coherent sentences in `zh-CN`, but the terminology I tend to use would look out of place to `zh-CN` users too.

I thought you could lock down certain strings as "approved strings" in Weblate so that only reviewers can edit them. For instance, the Plume AQI descriptions are locked down in `zh-TW`. If so, key weather terminology from a well-defined list (e.g. Beaufort wind forces or UV Index thresholds) should also be marked as "approved" once they are checked against verifiable sources (can be in comments). Meanwhile, other UI strings can remain open for revisions by everyone.

### Handling of HKO specific weather and waring strings
When I created the `hko_weather_text_*` and `hko_warning_text_*` strings in `strings.xml`, I was simply following your advice from the time I worked on Meteo AM. If you think that approach is not good for the HKO source, I'm fine moving them out of `strings.xml` and hardcoding them in the HKO source code instead. Just confirm your preference and I'll make the necessary changes:
- Keep both `hko_weather_text_*` strings and `hko_warning_text_*` strings in `strings.xml`
- Keep `hko_weather_text_*` strings in `strings.xml`; hardcode `hko_warning_text_*` strings
- Hardcode both `hko_weather_text_*` strings and `hko_warning_text_*` strings
- Other approach

**Comment by papjul** [comment]
I was just talking about adding `translatable="false"` attribute to the added strings. Then, add Indonesian, Hindi and Vietnamese in your PR as you won't be able to do it from Weblate since they are tagged as non translatable and Weblate ignores them ;)

**Comment by chunshek** [comment]
Oh yeah sure, that I can do! (I thought we only added `translatable="false"` for `@string/*` references.)
I'll add that to my to-do list after you've gone through the rest of the commit.

BTW, maybe a dumb question: if Weblate doesn't know a new string has been manually added to a particular translation file in Github, wouldn't the next commit from Weblate overwrite it and delete the string altogether?

**Comment by chunshek** [comment]
Hi @papjul. Is there anything else in the HKO code you would like me to change/fix?
If not, I'll sort out the resource strings as you suggested above later today. Thanks!

**Comment by papjul** [comment]
I didn't have time yet to review it, and won't have time today, so you can do it for sure ;)

**Comment by papjul** [comment]
Needs rebase ;)

**Comment by chunshek** [comment]
Done. Please check. :)

File: app/src/main/java/org/breezyweather/common/source/LocationPreset.kt [hunk 1]
```diff
@@ -64,7 +64,7 @@ enum class LocationPreset(
     // Asia
     // Do NOT set up other sources as only 中国 source is not rate-limited by the Great Firewall
     CHINA("china"),
-    HONG_KONG("openmeteo", alert = "wmosevereweather", normals = "accu"),
+    HONG_KONG("hko", alert = "hko", normals = "hko"),
```
Comment (papjul) [inline app/src/main/java/org/breezyweather/common/source/LocationPreset.kt:67]: Just `HONG_KONG("hko")` works

File: app/src/main/res/values-zh-rHK/strings.xml [hunk 2]
```diff
@@ -10,7 +10,7 @@
     <string name="air_quality_level_4">不健康</string>
     <string name="air_quality_level_5">很不健康</string>
     <string name="air_quality_pm25">PM2.5</string>
-    <string name="air_quality_pm25_voice">細懸浮微粒</string>
+    <string name="air_quality_pm25_voice">微細懸浮粒子</string>
```
Comment (papjul) [inline app/src/main/res/values-zh-rHK/strings.xml:13]: Is it normal other strings than `hko_` ones are changed?

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -819,6 +819,14 @@
...
     <string name="weather_kind_thunderstorm">Thunderstorm</string>
     <string name="weather_kind_wind">Wind</string>
+    <string name="weather_kind_mist">Mist</string>
+    <string name="weather_kind_mainly_cloudy">Mainly cloudy</string>
+    <string name="weather_kind_dry">Dry</string>
+    <string name="weather_kind_humid">Humid</string>
+    <string name="weather_kind_hot">Hot</string>
+    <string name="weather_kind_warm">Warm</string>
+    <string name="weather_kind_cool">Cool</string>
+    <string name="weather_kind_cold">Cold</string>
```
Comment (papjul) [inline app/src/main/res/values/strings.xml:829]: These weather kind don't exist in Breezy Weather. If they are HKO-specific, move them as `hko_` and `translatable="false"`.

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -894,6 +902,55 @@
     <string name="meteoam_weather_text_sand_storm" translatable="false">@string/common_weather_text_sand_storm</string>
     <!-- Example: Heavy rain and thunder -->
     <string name="metno_weather_text_andthunder">%s and thunder</string>
+    <!-- Weather texts for HKO -->
+    <string name="hko_weather_text_sunny" translatable="false">@string/common_weather_text_clear_sky</string>
+    <string name="hko_weather_text_sunny_periods">Sunny periods</string>
+    <string name="hko_weather_text_sunny_intervals">Sunny intervals</string>
+    <string name="hko_weather_text_sunny_periods_with_a_few_showers">Sunny periods with a few showers</string>
+    <string name="hko_weather_text_sunny_intervals_with_showers">Sunny intervals with showers</string>
```
Comment (papjul) [inline app/src/main/res/values/strings.xml:910]: Missing `translatable="false"` for a few strings

File: app/src/main/java/org/breezyweather/common/source/LocationPreset.kt (hunk 1 above)
Comment (chunshek) [reply app/src/main/java/org/breezyweather/common/source/LocationPreset.kt:67]: Fixed.

File: app/src/main/res/values-zh-rHK/strings.xml (hunk 2 above)
Comment (chunshek) [reply app/src/main/res/values-zh-rHK/strings.xml:13]: I've changed the word for "particulate matter" and "mold" to HK usage in a few strings. But I could revert these and edit them on Weblate instead if it's easier for your review. There are no usable air quality and pollen sources for HK now anyway.

File: app/src/main/res/values/strings.xml (hunk 3 above)
Comment (chunshek) [reply app/src/main/res/values/strings.xml:829]: Sure. I'll move these. I have a question and a comment though:

1. What's the difference between the `common_weather_text_` labels and the `weather_kind_` labels?
2. A few of these weather types will become shared with other sources. If I remember correctly, MGM (Türkiye) will share "mist", "hot" and "cold" with HKO. While "smoke" (currently coded for Meteo AM) will be used by MGM as well.

File: app/src/main/res/values/strings.xml (hunk 4 above)
Comment (chunshek) [reply app/src/main/res/values/strings.xml:910]: I thought people _may_ want to translate these, but I'll just lock them down as `translatable="false"` as you suggested.

**Comment by papjul** [comment]
Cleanup commit FYI: https://github.com/breezy-weather/breezy-weather/commit/d992c00f0c9efba8b2a2b623154fb70a77461bd5

**Comment by chunshek** [comment]
Thanks!
